lib/ecto/changeset.ex

```diff
@@ -3288,11 +3358,21 @@ defmodule Ecto.Changeset do
     required = Keyword.get(opts, :required, false)
 
     if required,
-      do: [{error_field, {message(opts, "can't be blank"), [validation: :required]}}],
+      do: [{error_field, message(opts, "can't be blank", validation: :required)}],
       else: []
   end
 
-  defp message(opts, key \\ :message, default) do
+  defp message(opts, key \\ :message, default, message_opts) do
+    case Keyword.get(opts, key, default) do
+      {message, extra_opts} when is_binary(message) and is_list(extra_opts) ->
+        {message, Keyword.merge(message_opts, extra_opts)}
```

Comment:
Should we let users override the ones being set by Ecto? I'm not sure if they are public and other libraries are depending on them.

Reply:
They are always public and I think overriding is fine. :)